ticketvote: Comittment addresses crash.

This crash was caused by a vote with a duration of 1 block where the vote ended before the commitment addresses cache finished populating.

```
2021-04-02 19:20:59.528 [INF] POLI: 127.0.0.1:16151 Plugin 'ticketvote' write cmd 'castballot' executed
2021-04-02 19:20:59.639 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 1000/5043
2021-04-02 19:21:00.722 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 1500/5043
2021-04-02 19:21:01.654 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 2000/5043
2021-04-02 19:21:03.359 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 2500/5043
2021-04-02 19:21:04.040 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 3000/5043
2021-04-02 19:21:04.756 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 3500/5043
2021-04-02 19:21:05.386 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 4000/5043
2021-04-02 19:21:06.076 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 4500/5043
2021-04-02 19:21:07.710 [DBG] PLUG: Get f8e609cc06e5765d commitment addrs 5000/5043
2021-04-02 19:21:08.155 [INF] POLI: 127.0.0.1:16151 POST /v2/pluginreads HTTP/1.1
2021-04-02 19:21:08.249 [DBG] PLUG: Quorum not met on f8e609cc06e5765d: votes cast 0, quorum 50
2021-04-02 19:21:08.250 [DBG] PLUG: Saved votes summary: f8e609cc06e5765d
2021-04-02 19:21:08.250 [DBG] PLUG: Active votes del f8e609cc06e5765d
2021-04-02 19:21:08.251 [DBG] PLUG: Vote inv update f8e609cc06e5765d to rejected
2021-04-02 19:21:08.252 [DBG] PLUG: Vote inv updated for block 654593
2021-04-02 19:21:08.340 [INF] POLI: 127.0.0.1:16151 POST /v2/pluginreads HTTP/1.1
2021-04-02 19:21:08.403 [INF] POLI: 127.0.0.1:21316 POST /v2/pluginreads HTTP/1.1
2021-04-02 19:21:08.436 [INF] POLI: 127.0.0.1:40519 POST /v2/records HTTP/1.1
panic: active vote not found f8e609cc06e5765d

goroutine 1406 [running]:
github.com/decred/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote.(*activeVotes).AddCommitmentAddrs(0xc000790b20, 0xc000cd0cd0, 0x10, 0xc0024558c0)
        /home/user/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go:219 +0x3ff
github.com/decred/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote.(*ticketVotePlugin).activeVotePopulateAddrs(0xc000697440, 0xc000cd0cd0, 0x10, 0x100000001, 0x3, 0x100000001, 0x32, 0xc001d82500, 0x2, 0x4, ...)
        /home/user/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go:300 +0x3b9
created by github.com/decred/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote.(*ticketVotePlugin).activeVotesAdd
        /home/user/politeia/politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go:314 +0xb8
```

--- a/politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go
+++ b/politeiad/backendv2/tstorebe/plugins/ticketvote/activevotes.go
@@ -6,7 +6,6 @@ package ticketvote
 
 import (
 	"encoding/hex"
-	"fmt"
 	"sync"
 
 	"github.com/decred/politeia/politeiad/plugins/ticketvote"
@@ -122,21 +121,23 @@ func (a *activeVotes) EligibleTickets(token []byte) map[string]struct{} {
 	return eligible
 }
 
-// VoteIsDuplicate returns whether the vote has already been cast. This
-// function will panic if the provided token does not correspond to a record in
-// the active votes cache.
-func (a *activeVotes) VoteIsDuplicate(token, ticket string) bool {
+// VoteIsDuplicate returns whether the vote has already been cast. The first
+// bool returned represents whether the record vote exists in the active votes
+// cache. The second bool returned represetns whether the ticket is a duplicate
+// vote.
+func (a *activeVotes) VoteIsDuplicate(token, ticket string) (bool, bool) {
 	a.RLock()
 	defer a.RUnlock()
 
 	av, ok := a.activeVotes[token]
 	if !ok {
-		// This should not happen
-		panic(fmt.Sprintf("active vote not found %v", token))
+		// Vote does not exist. Its possible that the vote
+		// ended while a ballot was being validated.
+		return false, false
 	}
 
 	_, isDup := av.CastVotes[ticket]
-	return isDup
+	return true, isDup
 }
 
 // CommitmentAddrs returns the largest comittment address for each of the
@@ -190,33 +191,34 @@ func (a *activeVotes) Tally(token string) map[string]uint32 {
 	return tally
 }
 
-// AddCastVote adds a cast ticket vote to the active votes cache. This function
-// will panic if the provided token does not correspond to a record in the
-// active votes cache.
+// AddCastVote adds a cast ticket vote to the active votes cache.
 func (a *activeVotes) AddCastVote(token, ticket, votebit string) {
 	a.Lock()
 	defer a.Unlock()
 
 	av, ok := a.activeVotes[token]
 	if !ok {
-		// This should not happen
-		panic(fmt.Sprintf("active vote not found %v", token))
+		// Vote does not exist. Its possible that the vote ended after
+		// the cast votes passed validation but before this cache was
+		// able to be populated. Log a warning and exit gracefully.
+		log.Warnf("AddCastVote: vote not found %v", token)
+		return
 	}
 
 	av.CastVotes[ticket] = votebit
 }
 
-// AddCommitmentAddrs adds commitment addresses to the cache for a record. This
-// function will panic if the provided token does not correspond to a record in
-// the active votes cache.
+// AddCommitmentAddrs adds commitment addresses to the cache for a record.
 func (a *activeVotes) AddCommitmentAddrs(token string, addrs map[string]commitmentAddr) {
 	a.Lock()
 	defer a.Unlock()
 
 	av, ok := a.activeVotes[token]
 	if !ok {
-		// This should not happen
-		panic(fmt.Sprintf("active vote not found %v", token))
+		// Vote does not exist. Its possible for the vote to end while
+		// in the middle of populating the commitment addresses cache.
+		// This is ok. Exit gracefully.
+		return
 	}
 
 	for ticket, v := range addrs {

--- a/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go
+++ b/politeiad/backendv2/tstorebe/plugins/ticketvote/cmds.go
@@ -1411,8 +1411,8 @@ func (p *ticketVotePlugin) cmdCastBallot(token []byte, payload string) (string, 
 	}
 
 	// Get the data that we need to validate the votes
-	voteDetails := p.activeVotes.VoteDetails(token)
 	eligible := p.activeVotes.EligibleTickets(token)
+	voteDetails := p.activeVotes.VoteDetails(token)
 	bestBlock, err := p.bestBlock()
 	if err != nil {
 		return "", err
@@ -1491,7 +1491,15 @@ func (p *ticketVotePlugin) cmdCastBallot(token []byte, payload string) (string, 
 		}
 
 		// Verify ticket has not already voted
-		if p.activeVotes.VoteIsDuplicate(v.Token, v.Ticket) {
+		isActive, isDup := p.activeVotes.VoteIsDuplicate(v.Token, v.Ticket)
+		if !isActive {
+			e := ticketvote.VoteErrorVoteStatusInvalid
+			receipts[k].Ticket = v.Ticket
+			receipts[k].ErrorCode = e
+			receipts[k].ErrorContext = fmt.Sprintf("%v: vote is "+
+				"not active", ticketvote.VoteErrors[e])
+		}
+		if isDup {
 			e := ticketvote.VoteErrorTicketAlreadyVoted
 			receipts[k].Ticket = v.Ticket
 			receipts[k].ErrorCode = e

